openwisp_controller/config/tasks.py

```diff
@@ -104,3 +104,59 @@ def trigger_vpn_server_endpoint(endpoint, auth_token, vpn_id):
             f'Response status code: {response.status_code}, '
             f'VPN Server UUID: {vpn_id}',
         )
+
+
+@shared_task(base=OpenwispCeleryTask)
+def change_devices_templates(instance_id, model_name, **kwargs):
+    def filter_backend_templates(templates, backend):
+        return filter(lambda template: template.backend == backend, templates)
+
+    def add_templates(device, templates, ignore_backend_filter=False):
+        if not ignore_backend_filter:
+            templates = filter_backend_templates(templates, device.config.backend)
+        device.config.templates.add(*templates)
+
+    def remove_templates(device, templates, ignore_backend_filter=False):
+        if not ignore_backend_filter:
+            templates = filter_backend_templates(templates, device.config.backend)
+        device.config.templates.remove(*templates)
+
+    Device = load_model('config', 'Device')
+    DeviceGroup = load_model('config', 'DeviceGroup')
+    Template = load_model('config', 'Template')
+    Config = load_model('config', 'Config')
+
+    if model_name == Device._meta.model_name:
+        device = Device.objects.get(pk=instance_id)
+        if not hasattr(device, 'config'):
+            return
+        old_group_id = kwargs.get('old_group_id')
+        group_id = kwargs.get('group_id')
+        group = DeviceGroup.objects.get(pk=group_id)
+        group_templates = group.templates.all()
+        if old_group_id:
+            old_group = DeviceGroup.objects.get(pk=old_group_id)
+            old_group_templates = old_group.templates.all()
+            remove_templates(device, old_group_templates)
+        add_templates(device, group_templates)
+
+    elif model_name == DeviceGroup._meta.model_name:
+        device_group = DeviceGroup.objects.get(id=instance_id)
+        templates = Template.objects.filter(pk__in=kwargs.get('templates'))
+        old_templates = Template.objects.filter(pk__in=kwargs.get('old_templates'))
+        for device in device_group.device_set.all():
+            if not hasattr(device, 'config'):
+                continue
+            remove_templates(device, old_templates)
+            add_templates(device, templates)
+
+    elif model_name == Config._meta.model_name:
+        config = Config.objects.get(pk=instance_id)
+        device_group = config.device.group
+        if not device_group:
+            return
+        templates = device_group.templates.filter(backend=kwargs.get('backend'))
+        old_templates = device_group.templates.filter(backend=kwargs.get('old_backend'))
+        ignore_backend_filter = True
+        remove_templates(config.device, old_templates, ignore_backend_filter)
+        add_templates(config.device, templates, ignore_backend_filter)
```

Comment:
I believe this logic should be moved in the models and split into different instance and/or class methods which are called here, this makes also unit testing easier because the test can target specific methods.

In general we shall put effort into ensuring that the core logic of the application is contained in the models.